vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1344 @@
+	entry, err := s.Get(ctx, publicKeysConfigPath+keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	var key jose.JSONWebKey
+	if err := entry.DecodeJSON(&key); err != nil {
+		return nil, err
+	}
+
+	return &key, nil
+}
+
+func listOIDCPublicKeys(ctx context.Context, s logical.Storage) ([]string, error) {
+	keys, err := s.List(ctx, publicKeysConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return keys, nil
+}
+
+func (i *IdentityStore) generatePublicJWKS(ctx context.Context, s logical.Storage) (*jose.JSONWebKeySet, error) {
+	if jwksRaw, ok := i.oidcCache.Get("jwks"); ok {
+		return jwksRaw.(*jose.JSONWebKeySet), nil
+	}
+
+	if _, err := i.expireOIDCPublicKeys(ctx, s); err != nil {
+		return nil, err
+	}
+
+	keyIDs, err := listOIDCPublicKeys(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	jwks := &jose.JSONWebKeySet{
+		Keys: make([]jose.JSONWebKey, 0, len(keyIDs)),
+	}
+
+	for _, keyID := range keyIDs {
+		key, err := loadOIDCPublicKey(ctx, s, keyID)
+		if err != nil {
+			return nil, err
+		}
+		jwks.Keys = append(jwks.Keys, *key)
+	}
+
+	i.oidcCache.SetDefault("jwks", jwks)
+
+	return jwks, nil
+}
+
+func (i *IdentityStore) expireOIDCPublicKeys(ctx context.Context, s logical.Storage) (time.Time, error) {
+	var didUpdate bool
+
+	// nextExpiration will be the soonest expiration time of all keys. Initialize
+	// here to a relatively distant time.
+	nextExpiration := time.Now().Add(24 * time.Hour)
+	now := time.Now()
+
+	publicKeyIDs, err := listOIDCPublicKeys(ctx, s)
+	if err != nil {
+		return now, err
+	}
+
+	namedKeys, err := s.List(ctx, namedKeyConfigPath)
+	if err != nil {
+		return now, err
+	}
+
+	usedKeys := make([]string, 0, 2*len(namedKeys))
+
+	for _, k := range namedKeys {
+		entry, err := s.Get(ctx, namedKeyConfigPath+k)
+		if err != nil {
+			return now, err
+		}
+
+		var key namedKey
+		if err := entry.DecodeJSON(&key); err != nil {
+			return now, err
+		}
+
+		// Remove any expired keys from the keyring.
+		keyRing := key.KeyRing
+		var keyringUpdated bool
+
+		for i := 0; i < len(keyRing); i++ {
+			k := keyRing[i]
+			if !k.ExpireAt.IsZero() && k.ExpireAt.Before(now) {
+				keyRing[i] = keyRing[len(keyRing)-1]
+				keyRing = keyRing[:len(keyRing)-1]
+
+				keyringUpdated = true
+				i--
+				continue
+			}
+
+			// Save a remaining key's next expiration if it is the earliest we've
+			// seen (for use by the periodicFunc for scheduling).
+			if !k.ExpireAt.IsZero() && k.ExpireAt.Before(nextExpiration) {
+				nextExpiration = k.ExpireAt
+			}
+
+			// Mark the KeyID as in use so it doesn't get deleted in the next step
+			usedKeys = append(usedKeys, k.KeyID)
+		}
+
+		// Persist any keyring updates if necessary
+		if keyringUpdated {
+			key.KeyRing = keyRing
+			entry, err := logical.StorageEntryJSON(entry.Key, key)
+			if err != nil {
+				i.Logger().Error("error updating key", "key", key.Name, "error", err)
+			}
+
+			if err := s.Put(ctx, entry); err != nil {
```

Comment:
Do we need locking between here and other functions that persist the keyring?

Reply:
Probably prudent